### DISCUSSION TITLE
How to automate deployment using HyperV?

### DISCUSSION THREAD

**Discussion by Flightkick** [siderolabs/talos#6018] [Q&A]
In order to get a development environment going I've been trying to automate deployment of a Talos (v1.1) cluster using HyperV.

**My environment:**
The toolstack I'm using for this is as follows:

- Windows 10 Pro 21H2
- [Hashicorp Packer](https://www.packer.io/) to build the HyperV box
- [Hashicorp Vagrant](https://www.vagrantup.com/) to deploy the box on the local HyperV hypervisor for the development environment*
- [Hashicorp Terraform](https://www.terraform.io/) for environment agnostic deployment, with the [`bmatcuk/vagrant`](https://github.com/bmatcuk/terraform-provider-vagrant) provider to deploy infrastructure on development environments.

_\* Sidenote: Using Vagrant comes with some challenges as it is opinionated in assuming an SSH connection will be available. I could probably work around that by disabling the communicator and using post-hooks to configure and finalize the deployment. But this is not an issue for now._

**The issue**
The main issue I'm facing is that, when Talos is deployed to HyperV, the VM won't advertise it's IP address to the hypervisor.
I found a [somewhat related post](https://superuser.com/a/1505724/509473) which explains that on Ubuntu, certain packages are necessary to make this work.
This behavior is the same whether the VM is deployed by Vagrant or manually using the [New-TalosVM Powershell module.](https://www.talos.dev/v1.1/talos-guides/install/virtualized-platforms/hyper-v/)
In case anyone would like to replicate, the original Powershell module didn't work for me as it was missing some configuration.
Check [my PR](https://github.com/nebula-it/New-TalosVM/pull/2) if the original module doesn't work for you.

I'm not that familiar with Talos Linux, but if I understand the docs correctly I would assume I could use [extension services](https://www.talos.dev/v1.1/advanced/extension-services/) for the purpose of installing such packages.

Unfortunately, and this is the crux; In order to apply this configuration, I would need to have the VM's IP address in order to communicate with it's API. Which is not available, because the VM doesn't report it's IP.

Of course I could connect to the serial console and find the IP manually but I would prefer an automated solution.
Any thoughts? Has anyone managed to get automated deployments using HyperV working?

**Comment by smira** [comment]
The system extension can be bundled into the Talos install image which solves chicken-and-egg problem here. 

https://www.talos.dev/v1.1/talos-guides/configuration/system-extensions/#building-a-talos-image-with-system-extensions

**Reply by Flightkick** [reply]
Thanks for your response! A system extension might do the trick!

Since the docs mention:
> System extensions are only activated during the installation or upgrade of Talos Linux.

> During the initial install (e.g. when PXE booting or booting from an ISO), Talos will pull down container images for system extensions, validate them, and include them into the Talos initramfs image. **System extensions will be activated on boot and overlaid on top of the Talos root filesystem.**

Am I correct in that this means that pulling, validation and inclusion in the image will only happen during install of Talos Linux but the system extensions will stay active during normal runtime? Or would I need to combine system extensions (during install) and extension services (during runtime)?

**Reply by smira** [reply]
System extensions modify Talos `initramfs` image, so there are ways to enable a system extension:

* if Talos boots from the disk image, system extensions might be pre-bundled, and as Talos is already "installed", no install will happen
* if Talos boots from ISO or PXE boots, system extension might be bundled in same way, but as install on disk happens, system extensions should be enabled in the machine config
* if Talos is upgraded, system extension should be in the machine config to keep it bundled

**Reply by Flightkick** [reply]
Alright, I'll check it out. Thanks!

**Reply by Flightkick** [reply]
I managed to progress a bit, although I've encountered some issues which I'll file once I'm done with building the ISO.

I have a Docker image with the system extension ready but I'm stuck on building the ISO.
Running the build command gives me the following output:
```
$ sudo make image-metal IMAGER_SYSTEM_EXTENSIONS="registry.io/flightkick/myimage:latest"

ghcr.io/siderolabs/imager:v1.2.0-alpha.1-21-ge97b9f6d3
v1.2.0-alpha.1-21-ge97b9f6d3: Pulling from siderolabs/imager
Digest: sha256:557974f4456e7cb0d306fadde257479d353b303f4aab7d90009e7bedae48bc18
Status: Image is up to date for ghcr.io/siderolabs/imager:v1.2.0-alpha.1-21-ge97b9f6d3
ghcr.io/siderolabs/imager:v1.2.0-alpha.1-21-ge97b9f6d3
Extracting registry.io/flightkick/myimage:latest...
2022/08/05 20:36:24 creating image for metal
2022/08/05 20:36:24 creating RAW disk
2022/08/05 20:36:24 attaching loopback device
2022/08/05 20:36:24 discovered system extensions:
2022/08/05 20:36:24 NAME                               VERSION   AUTHOR
2022/08/05 20:36:24 myimage                            0.0.1     Flightkick
2022/08/05 20:36:24 validating system extensions
2022/08/05 20:36:24 compressing system extensions
2022/08/05 20:36:24 creating system extensions initramfs archive and compressing it
10 blocks
(stdin): 724 B / 5120 B = 0.141
2022/08/05 20:36:24 creating new partition table on /dev/loop0
2022/08/05 20:36:24 logical/physical block size: 512/512
2022/08/05 20:36:24 minimum/optimal I/O size: 512/512
2022/08/05 20:36:24 partitioning /dev/loop0 - EFI "105 MB"
2022/08/05 20:36:24 created /dev/loop0p1 (EFI) size 204800 blocks
2022/08/05 20:36:24 partitioning /dev/loop0 - BIOS "1.0 MB"
2022/08/05 20:36:24 created /dev/loop0p2 (BIOS) size 2048 blocks
2022/08/05 20:36:24 partitioning /dev/loop0 - BOOT "1.0 GB"
2022/08/05 20:36:24 created /dev/loop0p3 (BOOT) size 2048000 blocks
2022/08/05 20:36:24 partitioning /dev/loop0 - META "1.0 MB"
2022/08/05 20:36:24 created /dev/loop0p4 (META) size 2048 blocks
2022/08/05 20:36:24 partitioning /dev/loop0 - STATE "105 MB"
2022/08/05 20:36:24 created /dev/loop0p5 (STATE) size 204800 blocks
2022/08/05 20:36:24 partitioning /dev/loop0 - EPHEMERAL "0 B"
2022/08/05 20:36:24 created /dev/loop0p6 (EPHEMERAL) size 86016 blocks
2022/08/05 20:36:24 formatting the partition "/dev/loop0p1" as "vfat" with label "EFI"
2022/08/05 20:36:24 zeroing out "/dev/loop0p2"
2022/08/05 20:36:24 formatting the partition "/dev/loop0p3" as "xfs" with label "BOOT"
2022/08/05 20:36:24 zeroing out "/dev/loop0p4"
2022/08/05 20:36:24 zeroing out "/dev/loop0p5"
2022/08/05 20:36:25 zeroing out "/dev/loop0p6"
2022/08/05 20:36:25 detaching loopback device
2022/08/05 20:36:25 error mounting "/dev/loop0p3": error mounting: 1 error(s) occurred:
        invalid argument

gzip: stdin: unexpected end of file
tar: Child returned status 1
tar: Error is not recoverable: exiting now
```

I've tried enabling debug output in the Makefile and passing `--build-arg=WITH_DEBUG=true` and `--build-arg=DEBUG=true` in the hope to get more valuable information about the 'invalid argument' response, without success.

Unfortunately I'm lacking in experience/knowledge of these loopback devices and their role in this build process.
I'm having some trouble identifying the issue, help is welcome!

**Reply by frezbo** [reply]
you'd need to create a buildx builder with extra permissions:

```bash
docker buildx create --driver docker-container --driver-opt network=host --name talos --buildkitd-flags '--allow-insecure-entitlement security.insecure' --use
```

**Reply by Flightkick** [reply]
That builder was already present (it was mentioned in the Makefile).
I am building inside a VM though, are there any special capabilities required maybe?

**Reply by frezbo** [reply]
it's possible the vm might not have loopback support? Try deleting the builder and re-creating with the permissions required

**Reply by Flightkick** [reply]
Loopback should work on `Linux ubuntu-18 4.15.0-58-generic #64-Ubuntu`, the `hashicorp/bionic64` box under HyperV.

> Try deleting the builder and re-creating with the permissions required

I forgot to mention, I did that already after your [reply](https://github.com/siderolabs/talos/discussions/6018#discussioncomment-3337009) to be sure I didn't have any previous builders with the wrong permissions.
These are my current builders:
```
$ sudo docker buildx ls
NAME/NODE DRIVER/ENDPOINT             STATUS   PLATFORMS
talos *   docker-container
  talos0  unix:///var/run/docker.sock inactive
default   docker
  default default                     running  linux/amd64, linux/386
```

I'm not sure if the output from the `mount` command is that sparse or if imager hides other essential information. The `/bin/installer` binary from the `ghcr.io/siderolabs/imager:v1.2.0-alpha.1-23-g4fd676c04` image doesn't seem to expose any kind of `--verbose` flags.
Am I correct that the Imager is closed source? I couldn't find any, this makes it a bit harder to figure out what's happening.

I'm going to submit the issues I've found so far which can be helpful if anyone cares to reproduce.
Thanks for your help so far, if you or anyone else got some more idea's I'd be happy to hear them 🙂

**Reply by Flightkick** [reply]
Issues reported:

- https://github.com/siderolabs/talos/issues/6029
- https://github.com/siderolabs/talos/issues/6030
- https://github.com/siderolabs/talos/issues/6031
- https://github.com/siderolabs/talos/issues/6032
- https://github.com/siderolabs/talos/issues/6033
